packages/ui/src/Modal.tsx

```diff
@@ -0,0 +1,166 @@
+import React, { FC, ReactNode } from 'react'
+import ReactModal, { Props as ReactModalProps } from 'react-modal'
+import cn from 'classnames'
+import { DataTestProp } from '@hazelcast/helpers'
+import { X } from 'react-feather'
+
+import { IconButton } from './IconButton'
+
+import styles from './Modal.module.scss'
+import { Button, ButtonProps, ButtonDisabledProps, ButtonAccessibleIconLeftProps } from './Button'
+
+/*
+ * What we do here, is pull the properties from Button and rename them. Unfortunately, we have to pull
+ * both part of the union separately, as picking the result discriminates the type and breaks the constrains of A | B.
+ *
+ * When Storybook supports tsc@4.1, this can be replaced with:
+ * type MapObjectUnion<T> = T extends Record<string, unknown> ? ModalActionMappedProps<T> : T
+ * type ModalActionMappedProps<T> = { [U in keyof T as `action${Capitalize<U>}`]: T[U] }
+ */
+type ModalActionDisabledProps =
+  | {
+      actionDisabledTooltip: ButtonDisabledProps['disabledTooltip']
+      actionDisabled: ButtonDisabledProps['disabled']
+    }
+  | {
+      actionDisabledTooltip?: never
+      actionDisabled?: never
+    }
+
+type ModalActionIconProps =
+  | {
+      actionIconLeft: ButtonAccessibleIconLeftProps['iconLeft']
+      actionIconLeftAriaLabel: ButtonAccessibleIconLeftProps['iconLeftAriaLabel']
+    }
+  | {
+      actionIconLeft?: never
+      actionIconLeftAriaLabel?: never
+    }
+
+type ModalActionCoreProps =
+  | {
+      // Note: Source of `onClick` is nullable property from ButtonHTMLAttributes<HTMLButtonElement>
+      onAction: NonNullable<ButtonProps['onClick']>
+      actionLabel: string
+    }
+  | { onAction?: never; actionLabel?: never }
+
+export type ModalActionProps = ModalActionCoreProps & {
+  actionKind?: ButtonProps['kind']
+} & ModalActionDisabledProps &
+  ModalActionIconProps
+
+export type ModalProps = {
+  closable?: boolean
+  children?: ReactNode
+  title: string
```

Comment:
I am not sure but I think we have cases where we need to support several different actions. Even if we don't I think it is a reasonable expectation.
I would suggest something like this
```suggestion
export type ModalProps = {
  closable?: boolean
  children?: ReactNode
  title: string
  actions: ButtonProps[]
```

Reply:
Fair point. Addressed in: https://github.com/hazelcast/frontend-shared/pull/148/commits/b7dca752a32d3db5c02bf509336aed687c0c26e7.